avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastSerdeCache.java

```diff
@@ -275,27 +276,84 @@ public static boolean isSupportedForFastSerializer(Schema.Type schemaType) {
    * @return generic-class aware avro {@link FastSerializer}
    */
   public FastSerializer<?> getFastGenericSerializer(Schema schema) {
-    String schemaKey = getSchemaKey(schema, schema);
-
-    FastSerializer<?> serializer = fastGenericRecordSerializersCache.get(schemaKey);
-    if (serializer == null) {
-      serializer =
-          fastGenericRecordSerializersCache.putIfAbsent(schemaKey, new FastSerializerWithAvroGenericImpl(schema));
-      if (serializer == null) {
-        serializer = fastGenericRecordSerializersCache.get(schemaKey);
-        CompletableFuture.supplyAsync(() -> buildGenericSerializer(schema), executor).thenAccept(s -> {
-          fastGenericRecordSerializersCache.put(schemaKey, s);
-        });
-      }
-    }
-    return serializer;
+    return getDeSerializer(schema, schema, fastGenericRecordSerializersCache,
+        FastSerializerWithAvroGenericImpl.class, "buildGenericSerializer");
   }
 
   private String getSchemaKey(Schema writerSchema, Schema readerSchema) {
     return String.valueOf(Math.abs(Utils.getSchemaFingerprint(writerSchema))) + Math.abs(
         Utils.getSchemaFingerprint(readerSchema));
   }
 
+  /**
+   * This function can get the total schema number limit for {@link FastSerdeCache}
+   *
+   * @return current schema number limit
+   */
+  public int getGeneratedFastClassNumLimit() {
+    return generatedFastClassNumLimit;
+  }
+
+  /**
+   * Set the total schema number limit for {@link FastSerdeCache}.
+   * It is the total limit for fast Generic/Specific De/Serializer
+   */
+  private void setGeneratedFastClassNumLimit(int limit) {
+    generatedFastClassNumLimit = limit;
+  }
+
+  /**
+   * This function is a template to generate if needed and return specific/generic-class aware avro de/serializer
+   *
+   * @param writerSchema writer schema
+   * @param readerSchema reader schema (for serializer, this can be null)
+   * @param cache cache for specific/generic de/serializer
+   * @param avroImplClass generator for regular avro de/serializer
+   * @param buildFastDeSerializerFuncName the name of helper function to build specific/generic de/serializer
+   *
+   * @return a fast specific/generic de/serializer
+   */
+  private <T> T getDeSerializer(Schema writerSchema, Schema readerSchema, Map<String, T> cache, Class<?> avroImplClass,
+      String buildFastDeSerializerFuncName){
+    String schemaKey = getSchemaKey(writerSchema, readerSchema);
+    T deserializer = cache.get(schemaKey);
+    if (deserializer == null) {
+      try {
+        Constructor cs = avroImplClass.getConstructor(Schema.class, Schema.class);
+        deserializer = cache.putIfAbsent(schemaKey, (T)cs.newInstance(writerSchema, readerSchema));
```

Comment:
putIfAbsent 1st constructs the argument (calling cs) and only then checks if the map already has a value.
computeIfAbsent would allow you to defer calling cs and hence would be slightly more efficient